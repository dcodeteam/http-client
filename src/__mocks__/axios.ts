const axios = {
  isCancel: jest.fn(x => x && x.isCancel),

  CancelTokens: [],
  CancelToken: {
    source: jest.fn(() => {
      const cancelToken = {
        token: {},
        cancel: jest.fn()
      };

      // @ts-ignore
      axios.CancelTokens.push(cancelToken);

      return cancelToken;
    })
  },

  instances: [],
  create: jest.fn(() => {
    const instance = {
      request: jest.fn(
        config =>
          new Promise((resolve, reject) => {
            if (config.url === "error") {
              const error = new Error("Http Error");

              // @ts-ignore
              error.config = config;

              reject(error);
            } else if (config.url === "cancel") {
              const error = new Error("Cancel Error");

              // @ts-ignore
              error.isCancel = true;

              reject(error);
            } else {
              resolve({ data: config });
            }
          })
      )
    };

    // @ts-ignore
    axios.instances.push(instance);

    return instance;
  })
};

export default axios;
