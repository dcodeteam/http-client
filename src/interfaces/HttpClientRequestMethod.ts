export type HttpClientRequestMethod = "GET" | "POST" | "PUT" | "DELETE";
