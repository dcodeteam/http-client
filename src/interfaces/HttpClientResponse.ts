export interface HttpClientResponse<T = any> {
  data: T;
  status: number;
  headers: object;
}
