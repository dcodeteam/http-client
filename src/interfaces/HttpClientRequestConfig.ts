import { HttpClientRequestMethod } from "./HttpClientRequestMethod";

export interface HttpClientRequestConfig {
  readonly method: HttpClientRequestMethod;

  readonly url: string;
  readonly pathParams?: object;
  readonly queryParams?: object;

  readonly data?: object;
  readonly headers?: object;

  readonly timeout?: number;
}
