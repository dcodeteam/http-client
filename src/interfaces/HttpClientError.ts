import { HttpClientRequestConfig } from "./HttpClientRequestConfig";
import { HttpClientResponse } from "./HttpClientResponse";

export interface HttpClientError extends Error {
  readonly code?: string;
  readonly response?: HttpClientResponse;
  readonly config: HttpClientRequestConfig;
}
