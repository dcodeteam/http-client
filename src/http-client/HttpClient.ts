import axios, { AxiosError, AxiosInstance, AxiosRequestConfig } from "axios";
import { Observable } from "rxjs";
import { retryWhen, scan, shareReplay } from "rxjs/operators";

import { HttpClientError } from "../interfaces/HttpClientError";
import {
  HttpClientFetchRequestConfig,
  HttpClientRequestConfig,
  HttpClientUpdateRequestConfig
} from "../interfaces/HttpClientRequestConfig";
import { HttpClientResponse } from "../interfaces/HttpClientResponse";
import { generatePath } from "../utils/GeneratePath";
import { createHttpClientError } from "../utils/HttpErrorUtils";

function isAxiosError(error: Error): error is AxiosError {
  return "config" in error;
}

export interface HttpClientOptions {
  readonly shouldRetry?: (
    info: {
      attempt: number;
      error: Error | HttpClientError;
      config: HttpClientRequestConfig;
    }
  ) => boolean;

  readonly errorInterceptor?: (error: Error | HttpClientError) => void;
  readonly requestInterceptor?: (config: HttpClientRequestConfig) => void;
  readonly responseInterceptor?: (
    config: HttpClientRequestConfig,
    response: HttpClientResponse
  ) => void;
}

/**
##### Initialization

```javascript
export function createHttpClient(logger) {
  return new HttpClient({
    requestInterceptor(config) {
      logger.logRequest(config);
    },
 *
    responseInterceptor(config, response) {
      logger.logResponse(config, response);
    },
 *
    errorInterceptor(error) {
      logger.logError(error);
    },
  });
}
```

##### Request
```javascript
const USER_URL = "/user/:userId";
const USER_COMMENTS_URL = "/user/:userId/comments";

export function fetchUser(httpClient, userId) {
  return httpClient.get(USER_URL, {
    pathParams: { userId }
  });
}

export function fetchUserPosts(httpClient, userId, page, limit) {
  return httpClient.get(USER_COMMENTS_URL, {
    url: USER_URL,
    pathParams: { userId },
    queryParams: { page, limit }
  });
}
```
 */
export class HttpClient {
  private readonly client: AxiosInstance;

  private readonly options: HttpClientOptions;

  public constructor(options: HttpClientOptions = {}) {
    this.options = options;
    this.client = axios.create();
  }

  public get<T>(
    url: string,
    options?: HttpClientFetchRequestConfig
  ): Observable<HttpClientResponse<T>> {
    return this.request<T>({ ...options, url, method: "GET" });
  }

  public post<T>(
    url: string,
    options?: HttpClientUpdateRequestConfig
  ): Observable<HttpClientResponse<T>> {
    return this.request<T>({ ...options, url, method: "POST" });
  }

  public put<T>(
    url: string,
    options?: HttpClientUpdateRequestConfig
  ): Observable<HttpClientResponse<T>> {
    return this.request<T>({ ...options, url, method: "PUT" });
  }

  public patch<T>(
    url: string,
    options?: HttpClientUpdateRequestConfig
  ): Observable<HttpClientResponse<T>> {
    return this.request<T>({ ...options, url, method: "PATCH" });
  }

  public delete<T>(
    url: string,
    options?: HttpClientFetchRequestConfig
  ): Observable<HttpClientResponse<T>> {
    return this.request<T>({ ...options, url, method: "DELETE" });
  }

  public request<T>(
    options: HttpClientRequestConfig
  ): Observable<HttpClientResponse<T>> {
    const {
      shouldRetry,
      errorInterceptor,
      requestInterceptor,
      responseInterceptor
    } = this.options;

    const stream = new Observable<HttpClientResponse<T>>(subscriber => {
      const cancelTokenSource = axios.CancelToken.source();

      if (requestInterceptor) {
        requestInterceptor(options);
      }

      const config: AxiosRequestConfig = {
        method: options.method,
        url: generatePath(options.url, options.pathParams),
        params: { ...options.queryParams },
        data: options.data,
        headers: { ...options.headers },
        timeout: options.timeout,
        cancelToken: cancelTokenSource.token
      };

      // Fix for react-native in Android devices.
      if (config.timeout == null || !isFinite(config.timeout)) {
        delete config.timeout;
      }

      let complete = false;

      this.client
        .request<T>(config)
        .then(x => {
          const response: HttpClientResponse<T> = {
            data: x.data,
            status: x.status,
            headers: x.headers
          };

          if (responseInterceptor) {
            responseInterceptor(options, response);
          }

          subscriber.next(response);
        })
        .catch(x => {
          if (axios.isCancel(x)) {
            return;
          }

          const error = !isAxiosError(x)
            ? x
            : createHttpClientError({
                code: x.code,
                config: options,
                message: x.message,
                response: x.response
              });

          if (errorInterceptor) {
            errorInterceptor(error);
          }

          subscriber.error(error);
        })
        .then(() => {
          complete = true;
          subscriber.complete();
        });

      return () => {
        if (!complete) {
          cancelTokenSource.cancel();
        }
      };
    });

    return !shouldRetry
      ? stream
      : stream.pipe(
          shareReplay(),
          retryWhen(errors =>
            errors.pipe(
              scan<Error, number>((attempt, error) => {
                if (shouldRetry({ error, config: options, attempt })) {
                  return attempt + 1;
                }

                throw error;
              }, 1)
            )
          )
        );
  }
}
