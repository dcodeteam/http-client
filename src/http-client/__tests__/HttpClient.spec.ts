import axios, { AxiosRequestConfig } from "axios";
import { Observable } from "rxjs";

import { HttpClientRequestConfig } from "../../interfaces/HttpClientRequestConfig";
import { HttpClientRequestMethod } from "../../interfaces/HttpClientRequestMethod";
import { isHttpClientError } from "../..";
import { HttpClient } from "../HttpClient";

jest.mock("axios");

describe("HttpClient", () => {
  const mockConfig: HttpClientRequestConfig = {
    method: "GET",

    url: "/foo/:bar",
    pathParams: { bar: "baz" },
    queryParams: { foo: "bar" },

    data: { foo: "bar" },
    headers: { foo: "bar" },

    timeout: 1000,
  };

  const httpMethods: HttpClientRequestMethod[] = [
    "GET",
    "POST",
    "PUT",
    "PATCH",
    "DELETE",
  ];

  describe("HttpClient#request", () => {
    it("should return observable", () => {
      const client = new HttpClient();

      expect(
        client.request({ url: mockConfig.url, method: mockConfig.method }),
      ).toBeInstanceOf(Observable);
    });

    it("should create new instance of axios client", () => {
      const spy = jest.spyOn(axios, "create");

      expect(spy).toHaveBeenCalledTimes(0);

      expect(() => new HttpClient()).not.toThrow();

      expect(spy).toHaveBeenCalledTimes(1);
    });

    it("should call 'axios.request'", async () => {
      const instance = axios.create();

      jest.spyOn(axios, "create").mockImplementationOnce(() => instance);

      const spy = jest.spyOn(instance, "request");

      const client = new HttpClient();
      const request = client.request({
        url: mockConfig.url,
        method: mockConfig.method,
      });

      await expect(request.toPromise()).resolves.toMatchSnapshot();

      expect(spy).toHaveBeenCalledTimes(1);
    });

    it("should accept 'method'", async () => {
      const client = new HttpClient();

      await Promise.all(
        httpMethods.map(async method => {
          const request = client.request<AxiosRequestConfig>({
            method,
            url: mockConfig.url,
          });

          const response = await request.toPromise();

          expect(response.data.method).toBe(method);
        }),
      );
    });

    it("should accept 'url'", async () => {
      const client = new HttpClient();

      const urls = ["/foo", "/bar", "/baz"];

      await Promise.all(
        urls.map(async url => {
          const request = client.request<AxiosRequestConfig>({
            url,
            method: mockConfig.method,
          });

          const response = await request.toPromise();

          expect(response.data.url).toBe(url);
        }),
      );
    });

    it("should accept 'pathParams'", async () => {
      const client = new HttpClient();

      const pathParams = ["foo", "bar", "baz"];

      await Promise.all(
        pathParams.map(async param => {
          const request = client.request<AxiosRequestConfig>({
            url: "/hello/:name",
            method: mockConfig.method,
            pathParams: { name: param },
          });

          const response = await request.toPromise();

          expect(response.data.url).toBe(`/hello/${param}`);
        }),
      );
    });

    it("should clone 'queryParams'", async () => {
      const client = new HttpClient();

      const queryList = [{ foo: "bar" }, { foo: "baz" }, { foo: "quoz" }];

      await Promise.all(
        queryList.map(async queryParams => {
          const request = client.request<AxiosRequestConfig>({
            queryParams,
            url: mockConfig.url,
            method: mockConfig.method,
          });

          const response = await request.toPromise();

          expect(response.data.params).toEqual(queryParams);
          expect(response.data.params).not.toBe(queryParams);
        }),
      );
    });

    it("should accept 'data'", async () => {
      const client = new HttpClient();

      const dataList = [{}, {}, {}];

      await Promise.all(
        dataList.map(async data => {
          const request = client.request<AxiosRequestConfig>({
            data,
            url: mockConfig.url,
            method: mockConfig.method,
          });

          const response = await request.toPromise();

          expect(response.data.data).toBe(data);
        }),
      );
    });

    it("should clone 'headers'", async () => {
      const client = new HttpClient();

      const headerList = [{ foo: "bar" }, { foo: "baz" }, { foo: "quoz" }];

      await Promise.all(
        headerList.map(async headers => {
          const request = client.request<AxiosRequestConfig>({
            headers,
            url: mockConfig.url,
            method: mockConfig.method,
          });

          const response = await request.toPromise();

          expect(response.data.headers).toEqual(headers);
          expect(response.data.headers).not.toBe(headers);
        }),
      );
    });

    it("should accept 'timeout'", async () => {
      const client = new HttpClient();

      const timeouts = [-100, 0, 100];

      await Promise.all(
        timeouts.map(async timeout => {
          const request = client.request<AxiosRequestConfig>({
            timeout,
            url: mockConfig.url,
            method: mockConfig.method,
          });

          const response = await request.toPromise();

          expect(response.data.timeout).toBe(timeout);
        }),
      );
    });

    it("should unset 'timeout' if it is not finite", async () => {
      const client = new HttpClient();
      const invalidValues = [undefined, Infinity, NaN];

      await Promise.all(
        invalidValues.map(async timeout => {
          const request = client.request<AxiosRequestConfig>({
            timeout,
            url: mockConfig.url,
            method: mockConfig.method,
          });

          const response = await request.toPromise();

          expect(response.data.timeout).toBeUndefined();
        }),
      );
    });

    it("should cancel request on observable close", () => {
      const token = axios.CancelToken.source();
      const spy = jest.spyOn(token, "cancel");

      jest
        .spyOn(axios.CancelToken, "source")
        .mockImplementationOnce(() => token);

      const client = new HttpClient();
      const request = client.request({
        url: "cancel",
        method: mockConfig.method,
      });

      const subscriber = request.subscribe();

      expect(spy).toHaveBeenCalledTimes(0);

      subscriber.unsubscribe();

      expect(spy).toHaveBeenCalledTimes(1);

      spy.mockRestore();
    });

    it("should throw HttpClientError on http errors", async () => {
      const client = new HttpClient();
      const request = client.request({
        url: "error",
        method: mockConfig.method,
      });

      try {
        await request.toPromise();
      } catch (e) {
        expect(e).toMatchSnapshot();
        expect(isHttpClientError(e)).toBe(true);
      }

      expect.assertions(2);
    });

    it("should throw Error on sync errors", async () => {
      const client = new HttpClient();
      const request = client.request({
        url: "/foo/:bar",
        pathParams: {},
        method: mockConfig.method,
      });

      try {
        await request.toPromise();
      } catch (e) {
        expect(e).toMatchSnapshot();
        expect(e).toBeInstanceOf(Error);
        expect(isHttpClientError(e)).toBe(false);
      }

      expect.assertions(3);
    });

    it("should call 'requestInterceptor' before request", async () => {
      const requestInterceptor = jest.fn();

      const options = { url: mockConfig.url, method: mockConfig.method };
      const client = new HttpClient({ requestInterceptor });
      const request = client.request(options);

      await request.toPromise();

      expect(requestInterceptor).toHaveBeenCalledTimes(1);
      expect(requestInterceptor).toBeCalledWith(options);
    });

    it("should call 'requestInterceptor' before request", async () => {
      const requestInterceptor = jest.fn();

      const config = {
        url: mockConfig.url,
        method: mockConfig.method,
      };
      const client = new HttpClient({ requestInterceptor });
      const request = client.request(config);

      await expect(request.toPromise()).resolves.toBeTruthy();

      expect(requestInterceptor).toHaveBeenCalledTimes(1);
      expect(requestInterceptor).toBeCalledWith(
        expect.objectContaining(config),
      );
    });

    it("should call 'responseInterceptor' before request", async () => {
      const responseInterceptor = jest.fn();

      const config = {
        url: mockConfig.url,
        method: mockConfig.method,
      };
      const client = new HttpClient({ responseInterceptor });
      const request = client.request(config);

      const response = await request.toPromise();

      expect(responseInterceptor).toHaveBeenCalledTimes(1);
      expect(responseInterceptor).toBeCalledWith(
        expect.objectContaining(config),
        response,
      );
    });

    it("should call 'errorInterceptor' on errors", async () => {
      const errorInterceptor = jest.fn();

      const client = new HttpClient({ errorInterceptor });
      const request = client.request({
        url: "error",
        method: mockConfig.method,
      });

      try {
        await request.toPromise();
      } catch (e) {
        expect(e).toMatchSnapshot();
        expect(errorInterceptor).lastCalledWith(e);
        expect(errorInterceptor).toHaveBeenCalledTimes(1);
      }

      expect.assertions(3);
    });

    it("should call 'shouldRetry' on errors", async () => {
      const shouldRetry = jest.fn(({ attempt }) => attempt < 3);

      try {
        const client = new HttpClient({ shouldRetry });
        const request = client.request({
          url: "error",
          method: mockConfig.method,
        });

        await request.toPromise();
      } catch (e) {
        expect(e).toMatchSnapshot();
        expect(shouldRetry).toHaveBeenCalledTimes(3);
        expect(shouldRetry.mock.calls).toEqual([
          [{ error: e, attempt: 1, config: e.config }],
          [{ error: e, attempt: 2, config: e.config }],
          [{ error: e, attempt: 3, config: e.config }],
        ]);
      }

      expect.assertions(3);
    });
  });

  httpMethods.forEach(x => {
    const method = x.toLowerCase();

    describe(`HttpClient#${method}`, () => {
      it("should call HttpClient#request", () => {
        const client = new HttpClient();
        const spy = jest.spyOn(client, "request");

        if (x === "GET") {
          expect(
            client.get(mockConfig.url, {
              timeout: mockConfig.timeout,
              headers: mockConfig.headers,
              pathParams: mockConfig.pathParams,
              queryParams: mockConfig.queryParams,
            }),
          ).toBeInstanceOf(Observable);
        } else if (x === "POST") {
          expect(
            client.post(mockConfig.url, {
              data: mockConfig.data,
              timeout: mockConfig.timeout,
              headers: mockConfig.headers,
              pathParams: mockConfig.pathParams,
              queryParams: mockConfig.queryParams,
            }),
          ).toBeInstanceOf(Observable);
        } else if (x === "PUT") {
          expect(
            client.put(mockConfig.url, {
              data: mockConfig.data,
              timeout: mockConfig.timeout,
              headers: mockConfig.headers,
              pathParams: mockConfig.pathParams,
              queryParams: mockConfig.queryParams,
            }),
          ).toBeInstanceOf(Observable);
        } else if (x === "PATCH") {
          expect(
            client.patch(mockConfig.url, {
              data: mockConfig.data,
              timeout: mockConfig.timeout,
              headers: mockConfig.headers,
              pathParams: mockConfig.pathParams,
              queryParams: mockConfig.queryParams,
            }),
          ).toBeInstanceOf(Observable);
        } else if (x === "DELETE") {
          expect(
            client.delete(mockConfig.url, {
              timeout: mockConfig.timeout,
              headers: mockConfig.headers,
              pathParams: mockConfig.pathParams,
              queryParams: mockConfig.queryParams,
            }),
          ).toBeInstanceOf(Observable);
        }

        expect(spy).toHaveBeenCalledTimes(1);
      });
    });
  });
});
