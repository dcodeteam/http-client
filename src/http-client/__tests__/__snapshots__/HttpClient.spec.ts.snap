// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HttpClient HttpClient#delete should call HttpClient#request 1`] = `
Array [
  Object {
    "headers": Object {
      "foo": "bar",
    },
    "method": "DELETE",
    "pathParams": Object {
      "bar": "baz",
    },
    "queryParams": Object {
      "foo": "bar",
    },
    "timeout": 1000,
    "url": "/foo/:bar",
  },
]
`;

exports[`HttpClient HttpClient#get should call HttpClient#request 1`] = `
Array [
  Object {
    "headers": Object {
      "foo": "bar",
    },
    "method": "GET",
    "pathParams": Object {
      "bar": "baz",
    },
    "queryParams": Object {
      "foo": "bar",
    },
    "timeout": 1000,
    "url": "/foo/:bar",
  },
]
`;

exports[`HttpClient HttpClient#post should call HttpClient#request 1`] = `
Array [
  Object {
    "data": Object {
      "foo": "bar",
    },
    "headers": Object {
      "foo": "bar",
    },
    "method": "POST",
    "pathParams": Object {
      "bar": "baz",
    },
    "queryParams": Object {
      "foo": "bar",
    },
    "timeout": 1000,
    "url": "/foo/:bar",
  },
]
`;

exports[`HttpClient HttpClient#put should call HttpClient#request 1`] = `
Array [
  Object {
    "data": Object {
      "foo": "bar",
    },
    "headers": Object {
      "foo": "bar",
    },
    "method": "PUT",
    "pathParams": Object {
      "bar": "baz",
    },
    "queryParams": Object {
      "foo": "bar",
    },
    "timeout": 1000,
    "url": "/foo/:bar",
  },
]
`;

exports[`HttpClient HttpClient#request should call 'axios.request' 1`] = `
Object {
  "data": Object {
    "cancelToken": Object {},
    "data": undefined,
    "headers": Object {},
    "method": "GET",
    "params": Object {},
    "url": "/foo/:bar",
  },
  "headers": undefined,
  "status": undefined,
}
`;

exports[`HttpClient HttpClient#request should cancel request on observable close 1`] = `
Array [
  Array [],
]
`;
